### DISCUSSION TITLE
Accessing the rays properties outside of the SystemLevelChannel class

### DISCUSSION THREAD

**Discussion by Elena-ID** [NVlabs/sionna#177] [Q&A] [answered]
Hi all,

I'm trying to access the angles (ZOA, AOA, ZOD, AOD) of the sampled rays outside of the SystemLevelChannel, and I haven't found a way to do it.
I want to use these angles (specifically the zenith angles) in a separate function, which I use after I initialize my UMi channel. I was thinking about adding my function to the ChannelCoefficientsGenerator class, but it might be overridden when I upgrade Sionna.
This might be a Python-related question more than Sionna-related, but I thought of asking here first.

Thanks in advance!

**Comment by jhoydis** [comment, marked as answer]
Hi,

I think that there is currently no way of getting the angles without hacking the source code. However, this could be very easy to do. 
You would simply need to have ``SystemLevelChannel.__call__()`` return the ``rays`` object.
That would boil down to replacing line 249 of [system_level_channel.py](https://github.com/NVlabs/sionna/blob/main/sionna/channel/tr38901/system_level_channel.py) with:
```
return h, delays, rays
```
The ``rays`` object has all of the properties you need, i.e., `['aoa',  'aod',  'delays',  'powers',  'xpr',  'zoa',  'zod']`.

We could consider adding a flag so that this object would be returned.

**Reply by Elena-ID** [reply]
That makes sense! 
Thanks for the help, Jakob!

**Reply by itsghaniz** [reply]
Hi, I want to do a similar thing to extract the individual ray data to develop a synthetic data set for an ML model. Has this feature been implemented in sionna as a function or is it still needed to be done through the source code? 
Thanks in advance

**Reply by jhoydis** [reply]
Hi @itsghaniz,

We have not added this feature yet.

**Reply by itsghaniz** [reply]
Thanks for the reply. As you have mentioned in the above discussion that hacking the source code can make this possible. I have done as instructed, but can you help me with what function to call to get this data? 
Thanks
